alf/trainers/policy_trainer.py

```diff
@@ -387,49 +333,67 @@ def _markdownify(paragraph):
-            max_to_keep=self._checkpoint_max_to_keep,
             algorithm=self._algorithm,
             metrics=metric_utils.MetricsGroup(self._driver.get_metrics(),
                                               'metrics'),
             global_step=global_step)
-        checkpointer.initialize_or_restore()
+        checkpointer.load()
         self._checkpointer = checkpointer
 
     def _save_checkpoint(self):
-        global_step = get_global_counter()
+        global_step = alf.summary.get_global_counter()
         self._checkpointer.save(global_step=global_step.numpy())
 
     def _eval(self):
-        global_step = get_global_counter()
-        with tf.summary.record_if(True):
-            eager_compute(
-                metrics=self._eval_metrics,
-                environment=self._eval_env,
-                state_spec=self._algorithm.predict_state_spec,
-                action_fn=lambda time_step, state: self._algorithm.predict(
-                    time_step=self._algorithm.transform_timestep(time_step),
-                    state=state,
-                    epsilon_greedy=self._config.epsilon_greedy),
-                num_episodes=self._num_eval_episodes,
-                step_metrics=self._driver.get_step_metrics(),
-                train_step=global_step,
-                summary_writer=self._eval_summary_writer,
-                summary_prefix="Metrics")
-            metric_utils.log_metrics(self._eval_metrics)
+        time_step = get_initial_time_step(self._eval_env)
+        policy_state = self._algorithm.get_initial_predict_state(
+            self._eval_env.batch_size)
+        episodes = 0
+        while episodes < self._num_eval_episodes:
+            time_step, policy_state = _step(
+                algortihm=self._algorithm,
+                env=self._eval_env,
+                time_ste=time_step,
+                policy_state=policy_state,
+                epsilon_greedy=self._config.epsilon_greedy,
+                metrics=self._eval_metrics)
+            if time_step.is_last():
+                episodes += 1
+        metric_utils.log_metrics(self._eval_metrics)
+
+
+def get_initial_time_step(env):
+    return common.zero_tensor_from_nested_spec(env.time_step_spec,
+                                               env.batch_size)
+
+
+def _step(algorithm, env, time_step, policy_state, epsilon_greedy, metrics):
+    policy_state = common.reset_state_if_necessary(
+        policy_state, algorithm.get_initial_predict_state(env.batch_size),
+        time_step.is_first())
+    transformed_time_step = algorithm.transform_timestep(time_step)
+    policy_step = algorithm.predict(transformed_time_step, policy_state,
+                                    epsilon_greedy)
+    next_time_step = env.step(policy_step.output)
+
+    exp = alf.data_structures.make_experience(time_step, policy_step,
+                                              policy_state)
+    for metric in metrics:
+        metric.observe(exp)
```

Comment:
metric(time_step) is the API defined here:
https://github.com/HorizonRobotics/alf/blob/torch_metrics/alf/metrics/metrics.py#L40